### ISSUE TITLE
Bug: Saving when File Path is Too Long

### ISSUE THREAD

**Issue by sahil48** [sumatrapdfreader/sumatrapdf#1016]
On windows, I am using Sumantra 3.1.2.

When I click File > Save As, and save a file in a directory, the file does not save because the directory+file path name is too long for the Windows file system. Adobe Reader gives me a warning when this happens. Sumantra does not, and I cannot figure out if the file was actually saved.

Please notify the user when a file cannot be saved for this reason.

**Comment by SumatraPeter** [comment]
Which version of Windows? 32 or 64-bit? 32 or 64-bit Sumatra? 

Here's a screenshot of 64-bit Sumatra 3.1.2 on Win10 1709:

![image](https://user-images.githubusercontent.com/15380332/39828894-4123b348-53da-11e8-92d3-f1a71eaf4e1c.png)

**Comment by sahil48** [comment]
It's in my program files folder and not program files (x86) folder so I think it is 64 bit. Can you also creating a bunch of folders and trying to save a shorter file name so that the path is exceeded. That's how I think it was failing.

**Comment by cweickhmann** [comment]
I have this problem, too, and it seems to be related to the difference between the length of the filename and the length of the complete path. Version and details are at the end of my post.

I cannot reproduce your screenshot, @SumatraPeter. When I go to C: for example, and type a long filename ( (255 times "a").pdf ), I get the following message (sorry, work PC... in German, but I think you agree it's misleading).
![sumatra-longfilename-error](https://user-images.githubusercontent.com/15192310/49948338-895ece80-fef3-11e8-93b3-4515afabf32c.png)

That being said: If I try to have the name longer than above (255+4=259) characters, it is cut off by the save dialogue. I guess that is a Windows "feature" as it appears in other programs as well. So, again, I guess the error message above may be caused by amiguous Windows error codes...

However, the thing I experience (and I guess @sahil48 too) is the following:
If the complete path (e.g. "C:\SomeFolder\.....\SomeSubfoldersLater\longfilename.pdf") is longer than 260 characters, SumatraPDF will close the dialogue with no prompt/error message, but the file is simply not saved.

Actually, Microsoft has introduced a notation for long filenames, starting with "\\?\", which might solve that issue.

DETAILS:
* SumatraPDF Version 3.1.2 64bit
* OS: Windows 10  V 1803 (build 17134.471)

**Comment by GitHubRulesOK** [comment]
FWIW Adobe gives no warning when a file is too long to open and just sits there I may be wrong but I don't remember it always telling when saving either. but without firing up the boilers on the old steam driven pc I I haven't got acrobat here to test at present.

**Comment by cweickhmann** [comment]
Nope, @GitHubRulesOK , Adobe Reader DC at least gives you feedback when the overall path length is exceeding 260 characters.

![adobe-long-filename-error](https://user-images.githubusercontent.com/15192310/49952652-a1871b80-fefc-11e8-89bb-1d0c5ea4ce18.png)

(Sorry, German interface again. Reads: "The Path is too long. Enter a shorter filename or save the file in a location with a shorter path.")

**Comment by SumatraPeter** [comment]
>I cannot reproduce your screenshot, @SumatraPeter. When I go to C: for example, and type a long filename ( (255 times "a").pdf ), I get the following message

>That being said: If I try to have the name longer than above (255+4=259) characters, it is cut off by the save dialogue. I guess that is a Windows "feature" as it appears in other programs as well. So, again, I guess the error message above may be caused by amiguous Windows error codes...

@cweickhmann: Looks like a Windows update/upgrade wisely put in place a length restriction for the file name field in the Save dialog, so I can't reproduce my earlier screenshot either with 64-bit Sumatra 3.1.2 on Win10 1803:

![image](https://user-images.githubusercontent.com/15380332/49957675-8fbe6c00-ff2e-11e8-9963-d9a1c54bc406.png)

The silent failure bug while attempting to save in an existing folder such that the full path to the saved file would be longer than max path still exists in the latest pre-release v11063.

**Comment by cweickhmann** [comment]
@SumatraPeter , so this boils down to a Windows bug?
Any idea how e.g. Acrobat circumvents this?

(I mean, for my specific workflow it's not a big deal. I save the PDF and then manually add the filename to a JabRef archive. When the file wasn't saved properly, I find out because JabRef tells me that the file isn't there. But I guess it's catastrophic for many other users. Especially when they have downloaded a large PDF to a temp folder and after closing the browser, the temp file is gone.)

**Comment by SumatraPeter** [comment]
@cweickhmann: No, it's a Windows limitation that Sumatra can now bypass AFAIK in [newer versions of Win10](https://www.howtogeek.com/266621/how-to-make-windows-10-accept-file-paths-over-260-characters/), or it can at the very least check for so as to throw an error.

That's _my_ understanding of the issue. The developer may have a different take on it.
